btree/btree.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import "fmt"
+
+type Node struct {
+	degree   int
+	keys     []int
+	isLeaf   bool
+	children []*Node
+}
+
+func NewNode(degree int) *Node {
+	n := &Node{}
+	n.isLeaf = true
+	n.degree = degree
+	return n
+}
+
+func (n *Node) InsertInt(idx int, val int) {
+	n.keys = append(n.keys, 0)
+	copy(n.keys[idx+1:], n.keys[idx:])
+	n.keys[idx] = val
+}
+
+func (n *Node) InsertNode(idx int, item *Node) {
+	n.children = append(n.children, nil)
+	copy(n.children[idx+1:], n.children[idx:])
+	n.children[idx] = item
+}
+
+func (n *Node) locateLocally(x int) int {
+	for k := range n.keys {
+		if x <= n.keys[k] {
+			return k
+		}
+	}
+	return len(n.keys)
+}
+
+func (n *Node) insertRec(x int) (k int, m *Node) {
+	i := n.locateLocally(x)
+	if n.isLeaf {
+		n.InsertInt(i, x)
+	} else {
+		s, l := n.children[i].insertRec(x)
+		if l != nil {
+			n.InsertInt(i, s)
+			n.InsertNode(i, l)
+
+		}
+	}
+	if len(n.keys) < n.degree {
+		k = 0
+		m = nil
+	} else {
+		mid := len(n.keys) / 2
+		k = n.keys[mid]
+		m = NewNode(n.degree)
+		m.isLeaf = n.isLeaf
+		m.keys = make([]int, mid)
+		copy(m.keys, n.keys[:mid])
+		n.keys = n.keys[mid+1:]
+		if !n.isLeaf {
+			m.children = make([]*Node, mid+1)
+			copy(m.children, n.children[:mid+1])
+			n.children = n.children[mid+1:]
+		}
+	}
+	return
+}
+
+type BTree struct {
+	top    *Node
+	degree int
+}
+
+func NewTree(degree int) *BTree {
+	top := NewNode(degree)
+	t := &BTree{}
+	t.top = top
+	t.degree = degree
+	return t
+}
+
+func (t *BTree) insert(x int) {
```

Comment:
```suggestion
func (t *BTree) Insert(x int) {
```
このメソッドは公開しないとまずい気がします

Reply:
全く気にしてなかったので修正しました。ありがとうございます。